tofu/data/_core.py

```diff
@@ -2418,71 +2418,176 @@ def _checkformat_dtrm(dtime=None, dradius=None, dmesh=None,
                         msg += "    - Expected: %s\n"%str(dkok[dk]['ndim'])
                         msg += "    - Provided: %s"%str(dd[dk][k0]['data'].ndim)
                         raise Exception(msg)
+
+                # mesh
                 if dk == 'dmesh':
-                    dd[dk][k0]['nodes'] = np.atleast_2d(v0['nodes']).astype(float)
-                    dd[dk][k0]['faces'] = np.atleast_2d(v0['faces']).astype(int)
-                    nnodes = dd[dk][k0]['nodes'].shape[0]
-                    nfaces = dd[dk][k0]['faces'].shape[0]
-
-                    # Test for duplicates
-                    nodesu = np.unique(dd[dk][k0]['nodes'], axis=0)
-                    facesu = np.unique(dd[dk][k0]['faces'], axis=0)
-                    lc = [nodesu.shape[0] != nnodes,
-                          facesu.shape[0] != nfaces]
-                    if any(lc):
-                        msg = "Non-valid mesh %s[%s]:\n"%(dk,k0)
-                        if lc[0]:
-                            msg += "  Duplicate nodes: %s\n"%str(nnodes - nodesu.shape[0])
-                            msg += "    - nodes.shape: %s\n"%str(dd[dk][k0]['nodes'].shape)
-                            msg += "    - unique nodes.shape: %s\n"%str(nodesu.shape)
-                        if lc[1]:
-                            msg += "  Duplicate faces: %s\n"%str(nfaces - facesu.shape[0])
-                            msg += "    - faces.shape: %s\n"%str(dd[dk][k0]['faces'].shape)
-                            msg += "    - unique faces.shape: %s"%str(facesu.shape)
-                        raise Exception(msg)
+                    assert v0['type'] in ['rect', 'tri', 'quadtri'], v0['type']
```

Comment:
I thought you wanted to get rid of "assert" statements?

Reply:
You're right, I replace it with if / raise